tests/ci/check_so_interface.py

```diff
@@ -276,68 +303,94 @@
-            raise NotImplementedError("This code doesn't currently have the functionality to support varargs, " + name + " has vararg")
-        if has_kwargs == True:
-            raise NotImplementedError("This code doesn't currently have the functionality to support kwargs, " + name + " has kwarg")
+def compare_args(node, intdict):
+    """
+    Uses intdict to compare the expected 'args', 'defaults', and 'kwargs' 
+    values for certain frozen function specified by node.name to those 
+    values extracted from node, printing feedback on detected differences 
+    and returning false or returning true if no differences detected.
 
-        # Compare arguments to specified arguments in interface dict
-        if found_args != spec_args:
-            print(f"Mismatch in arguments for function {name}")
-            print(f"Found: {found_args}")
-            print(f"Expected: {spec_args}")
-            return False
-        
-        # Compare default values to specified defaults in interface dict
-        if found_defaults != spec_defaults:
-            print(f"Mismatch in default values for function {name}")
-            print(f"Found: {found_defaults}")
-            print(f"Expected: {spec_defaults}")
-            return False
-    #Check class
-    #elif isinstance(node, ast.ClassDef):
+    Args
+    ---
+    node : an AST tree node
+        Node from the AST tree that pairs with a 
+        given function from the interface dictionary.
+    intdict : A dictionary
+        Dictionary that contains values for 'args',
+        'defaults' and 'kwargs' for a given function
 
+    Returns
+    -------
+    Boolean, or raises an error
+    """
+    #getting specified args and defaults from interface
+    spec_args = intdict['args']
+    spec_defaults = intdict['defaults']
+    # Extract function arguments from the AST node
+    found_args = [arg.arg for arg in node.args.args]
+    found_defaults = [ast.unparse(d) if d else None for d in node.args.defaults]
+    has_varargs = node.args.vararg is not None #check for varargs
+    has_kwargs = node.args.kwarg is not None #check for kwargs
+    #Check for varargs
+    if has_varargs:
+        raise NotImplementedError("This code doesn't currently have the functionality to support varargs, " + (node.name) + " has vararg")
+    #If kwargs detected but kwargs not expected
+    if has_kwargs and intdict.get('kwarg') == None:
+        raise Exception("Kwarg expected to be found in "+ (node.name) + " but kwarg not found")
+    #If kwargs not detected but kwargs expected
+    if not has_kwargs and not intdict.get('kwarg') == None:
+        raise Exception("Kwarg not expected to be found in " + (node.name) + " but kwarg detected")
+    # Compare arguments to specified arguments in interface dict
+    if found_args != spec_args:
+        print(f"Mismatch in arguments for function {node.name}")
+        print(f"Found: {found_args}")
+        print(f"Expected: {spec_args}")
+        return "incorrectargs"
```

Comment:
This works for sure, but it could be nicer if there was a more consistent behaviour in how the function behaved on success/failure.

Right now it can fail either by returning `False`, a `str`, or raising an `Exception`. This is not explained in the docstring. From a user's point of view it would be a lot clearer if there was just one behaviour to expect.

Given you would like to pass on some information about the failure, the best solution might be to raise an `Exception`. You already do that for the `kwargs`/`varargs` case, and you could just change these cases to return an `Exception` that contains the message you are currently printing.

See below for suggestions on handling the exceptions.